### PR TITLE
Implement max_executors_per_slave job conf setting

### PR DESCRIPTION
**Pull request by josephharrington**
We currently allow users to specify a "max_executors" setting on their
jobs. This is useful to prevent one job from "overparallelizing" and
monopolizing an entire cluster.

This change implements an additional job config setting called
"max_executors_per_slave". This is useful, for example, on jobs which
*can* be parallelized across multiple machines, but *cannot* be
parallelized with multiple processes on a single machine. (That problem
can now be solved by setting max_executors_per_slave to 1.)

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @josephharrington has signed the CLA. Thanks for the pull request!

File: app/master/build.py [hunk 1]
```diff
@@ -43,14 +43,16 @@ def __init__(self, build_request):
         self._slaves_allocated = []
         self._num_executors_allocated = 0
         self._num_executors_in_use = 0
+
         self._max_executors = float('inf')
-        self._build_completion_lock = Lock()
```
Comment (josephharrington) [inline app/master/build.py]: This is already defined on line 42 above.

File: app/master/build.py [hunk 2]
```diff
@@ -43,14 +43,16 @@ def __init__(self, build_request):
...
+        self._max_executors_per_slave = float('inf')
 
         self._all_subjobs_by_id = {}
         self._unstarted_subjobs = None
         self._finished_subjobs = None
         self._postbuild_tasks_are_finished = False
         self._teardowns_finished = False
+        self._timing_file_path = None
```
Comment (josephharrington) [inline app/master/build.py]: Random unrelated cleanup -- defining all attributes in constructor.

File: examples/directory job/clusterrunner.yaml [hunk 3]
```diff
@@ -1,4 +1,6 @@
 SampleJob:
+    max_executors: 2
+    max_executors_per_slave: 5
```
Comment (josephharrington) [inline examples/directory job/clusterrunner.yaml]: Just adding these here for basic reference purposes. This doesn't really do anything. :smile_cat:.

File: test/unit/master/test_build.py [hunk 4]
```diff
@@ -31,27 +33,54 @@ def test_allocate_slave_calls_slave_setup(self):
...
+        expected_current_num_executors_used = 0
+        for i in range(len(mock_slaves)):
+            build.begin_subjob_executions_on_slave(mock_slaves[i])
+            expected_current_num_executors_used += max_executors_per_slave
+            self.assertEqual(
+                build.execute_next_subjob_on_slave.call_count, expected_current_num_executors_used,
+                'After allocating {} slaves, build with max_executors_per_slave set to {} should only be using {} '
+                'executors.'.format(i + 1, max_executors_per_slave, expected_current_num_executors_used))
+
+        self.assertEqual(build.execute_next_subjob_on_slave.call_count, expected_total_num_executors_used,
```
Comment (josephharrington) [inline test/unit/master/test_build.py]: This last assertion is redundant, but I think it makes the test clearer. 

Just realized I need to fix the message for this assertion though. Will do.

**Comment by drobertduke** [comment]
great! :8ball:
